File: Harvest/ViewModels/ChatViewModel.swift
import Foundation
import Observation
import Realtime

@Observable
final class ChatViewModel {
    var messages: [Message] = []
    var partnerProfile: UserProfile?
    var isLoading = false
    var messageText = "" {
        didSet {
            if !messageText.isEmpty {
                scheduleTypingIndicator()
            }
        }
    }
    var error: String?

    // Typing indicator
    var isPartnerTyping = false
    private var typingChannel: RealtimeChannelV2?
    private var typingDebounceTask: Task<Void, Never>?
    private var typingDismissTask: Task<Void, Never>?
    private var activeConversationId = ""
    private var activeUserId = ""
    private var activeMatchId = ""
    private var activePartnerUserId = ""

    // Mindful messaging
    var mindfulAnalysis: MindfulMessagingService.MindfulAnalysis?
    var showMindfulWarning = false
    private var pendingMessageText = ""
    private var pendingConversationId = ""
    private var pendingSenderId = ""

    // Safety
    var safetyAnalysis: SafetyAnalysis?
    var safetyWarning: String?
    var showReadyToMoveGate = false
    var isReadyToMove = false
    var readyToMoveReason: String?
    var readyToMoveActionMessage: String?
    var showReadyToMoveActionAlert = false

    // Report/Block/Unmatch
    var showReportSheet = false
    var showBlockAlert = false
    var showUnmatchAlert = false

    private let chatService = ChatService()
    private let profileService = ProfileService()
    private let mindfulService = MindfulMessagingService()
    private let safetyService = SafetyAnalysisService()
    private let matchService = MatchService()
    private var channel: RealtimeChannelV2?

    func loadMessages(conversationId: String) async {
        isLoading = true
        defer { isLoading = false }

        do {
            messages = try await chatService.getMessages(conversationId: conversationId)
        } catch {
            self.error = error.localizedDescription
        }
    }

    func loadPartnerProfile(userId: String) async {
        do {
            partnerProfile = try await profileService.getProfile(userId: userId)
        } catch {
            // Non-critical
        }
    }

    func loadSafetyAnalysis(matchId: String, userId: String, otherUserId: String) async {
        do {
            activeMatchId = matchId
            activeUserId = userId
            activePartnerUserId = otherUserId

            safetyAnalysis = try await safetyService.getOrCreateAnalysis(
                matchId: matchId,
                userId: userId,
                otherUserId: otherUserId
            )

            let hasPartnerHistory = messages.contains {
                !$0.isSentBy(userId) && !($0.content?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ?? true)
            }

            if hasPartnerHistory, safetyAnalysis?.redFlagCount == 0, !activeConversationId.isEmpty {
                safetyAnalysis = try await safetyService.analyzeConversationHistory(
                    conversationId: activeConversationId,
                    matchId: matchId,
                    userId: userId,
                    otherUserId: otherUserId
                )
            }

            refreshSafetyWarning()
            await refreshReadyToMoveStatus()
        } catch {
            // Non-critical
        }
    }

    func sendMessage(conversationId: String, senderId: String) async {
        let text = messageText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else { return }

        // Mindful messaging check
        if mindfulService.isEnabled {
            let analysis = await mindfulService.analyzeMessage(text)
            if analysis.needsReview {
                mindfulAnalysis = analysis
                pendingMessageText = text
                pendingConversationId = conversationId
                pendingSenderId = senderId
                showMindfulWarning = true
                return
            }
        }

        await performSend(text: text, conversationId: conversationId, senderId: senderId)
    }

    func confirmSendDespiteWarning() async {
        showMindfulWarning = false
        mindfulAnalysis = nil
        await performSend(
            text: pendingMessageText,
            conversationId: pendingConversationId,
            senderId: pendingSenderId
        )
    }

    func dismissMindfulWarning() {
        showMindfulWarning = false
        mindfulAnalysis = nil
    }

    private func performSend(text: String, conversationId: String, senderId: String) async {
        messageText = ""

        do {
            if let message = try await chatService.sendMessage(
                conversationId: conversationId,
                senderId: senderId,
                content: text
            ) {
                // Only add if not already added by realtime
                if !messages.contains(where: { $0.id == message.id }) {
                    messages.append(message)
                }
            }
        } catch {
            self.error = error.localizedDescription
            messageText = text // Restore on failure
        }
    }

    func subscribeToRealtime(conversationId: String) {
        channel = chatService.subscribeToMessages(conversationId: conversationId) { [weak self] message in
            Task { @MainActor [weak self] in
                guard let self else { return }
                if !self.messages.contains(where: { $0.id == message.id }) {
                    self.messages.append(message)
                }
                await self.analyzeMessageForSafetyIfNeeded(message)
            }
        }
    }

    func subscribeToTyping(conversationId: String, currentUserId: String) {
        activeConversationId = conversationId
        activeUserId = currentUserId
        typingChannel = chatService.subscribeToTyping(conversationId: conversationId) { [weak self] userId in
            Task { @MainActor [weak self] in
                guard let self, userId != currentUserId else { return }
                self.isPartnerTyping = true
                self.typingDismissTask?.cancel()
                self.typingDismissTask = Task {
                    try? await Task.sleep(for: .seconds(3))
                    guard !Task.isCancelled else { return }
                    self.isPartnerTyping = false
                }
            }
        }
    }

    private func scheduleTypingIndicator() {
        typingDebounceTask?.cancel()
        typingDebounceTask = Task {
            try? await Task.sleep(for: .milliseconds(500))
            guard !Task.isCancelled else { return }
            if !activeConversationId.isEmpty, !activeUserId.isEmpty {
                await chatService.sendTypingIndicator(
                    conversationId: activeConversationId,
                    userId: activeUserId
                )
            }
        }
    }

    func markMessagesAsRead(currentUserId: String) async {
        let unreadMessages = messages.filter { !$0.isRead && !$0.isSentBy(currentUserId) }
        for message in unreadMessages {
            do {
                try await chatService.markAsRead(messageId: message.id)
                if let index = messages.firstIndex(where: { $0.id == message.id }) {
                    messages[index].isRead = true
                }
            } catch {
                // Non-critical
            }
        }
    }

    func unsubscribe() {
        if let channel {
            chatService.unsubscribe(channel: channel)
        }
        channel = nil
        if let typingChannel {
            chatService.unsubscribe(channel: typingChannel)
        }
        typingChannel = nil
        typingDebounceTask?.cancel()
        typingDismissTask?.cancel()
        activeMatchId = ""
        activePartnerUserId = ""
    }

    func ensureConversation(matchId: String, user1Id: String, user2Id: String) async -> String? {
        do {
            return try await chatService.ensureConversation(
                matchId: matchId,
                user1Id: user1Id,
                user2Id: user2Id
            )
        } catch {
            self.error = error.localizedDescription
            return nil
        }
    }

    // MARK: - Report/Block/Unmatch

    func reportUser(reporterId: String, reportedUserId: String, category: String, description: String) async {
        do {
            try await matchService.reportUser(
                reporterId: reporterId,
                reportedUserId: reportedUserId,
                category: category,
                description: description
            )
        } catch {
            self.error = error.localizedDescription
        }
    }

    func blockUser(userId: String, blockedUserId: String) async -> Bool {
        do {
            try await matchService.blockUser(userId: userId, blockedUserId: blockedUserId)
            return true
        } catch {
            self.error = error.localizedDescription
            return false
        }
    }

    func unmatchUser(matchId: String) async -> Bool {
        do {
            try await matchService.unmatchUser(matchId: matchId)
            return true
        } catch {
            self.error = error.localizedDescription
            return false
        }
    }

    // MARK: - Retroactive Safety Analysis

    func runRetroactiveSafetyAnalysis(conversationId: String, matchId: String, userId: String, otherUserId: String) async {
        do {
            let analysis = try await safetyService.analyzeConversationHistory(
                conversationId: conversationId,
                matchId: matchId,
                userId: userId,
                otherUserId: otherUserId
            )
            safetyAnalysis = analysis
            refreshSafetyWarning()
            await refreshReadyToMoveStatus()
            print("Retroactive analysis complete. Safety score: \(analysis.safetyScore), Red flags: \(analysis.redFlagCount)")
        } catch {
            print("Error running retroactive safety analysis: \(error)")
            self.error = "Failed to analyze conversation history"
        }
    }

    @MainActor
    private func analyzeMessageForSafetyIfNeeded(_ message: Message) async {
        guard
            let analysis = safetyAnalysis,
            !activeMatchId.isEmpty,
            !activeUserId.isEmpty,
            !activePartnerUserId.isEmpty,
            !message.isSentBy(activeUserId),
            let content = message.content?.trimmingCharacters(in: .whitespacesAndNewlines),
            !content.isEmpty
        else {
            return
        }

        do {
            _ = try await safetyService.analyzeMessage(content, analysisId: analysis.id)
            safetyAnalysis = try await safetyService.getOrCreateAnalysis(
                matchId: activeMatchId,
                userId: activeUserId,
                otherUserId: activePartnerUserId
            )
            refreshSafetyWarning()
            await refreshReadyToMoveStatus()
        } catch {
            print("Warning: Failed to analyze message for safety: \(error)")
        }
    }

    func presentReadyToMoveGate() async {
        await refreshReadyToMoveStatus()
        showReadyToMoveGate = true
    }

    func markPreferredContactShared() async {
        guard isReadyToMove, let analysis = safetyAnalysis else { return }

        do {
            try await safetyService.recordReadyToMoveDecision(
                analysisId: analysis.id,
                userId: activeUserId,
                approved: true,
                contactShared: true,
                contactMethod: "social"
            )
            readyToMoveActionMessage = "You're clear to share your preferred contact details in the chat."
            showReadyToMoveActionAlert = true
            showReadyToMoveGate = false
        } catch {
            readyToMoveActionMessage = "We couldn't record your sharing action. Please try again."
            showReadyToMoveActionAlert = true
        }
    }

    private func refreshSafetyWarning() {
        if let score = safetyAnalysis?.safetyScore, score < 50 {
            safetyWarning = "Safety concern detected. Be cautious in this conversation."
        } else {
            safetyWarning = nil
        }
    }

    private func refreshReadyToMoveStatus() async {
        guard var analysis = safetyAnalysis else {
            isReadyToMove = false
            readyToMoveReason = "No safety analysis available yet."
            return
        }

        do {
            if !activeMatchId.isEmpty, !activeUserId.isEmpty, !activePartnerUserId.isEmpty {
                analysis = try await safetyService.getOrCreateAnalysis(
                    matchId: activeMatchId,
                    userId: activeUserId,
                    otherUserId: activePartnerUserId
                )
                safetyAnalysis = analysis
            }

            let result = try await safetyService.isReadyToMove(analysisId: analysis.id)
            isReadyToMove = result.ready
            readyToMoveReason = result.reason
        } catch {
            isReadyToMove = false
            readyToMoveReason = "Couldn't determine whether this conversation is ready yet."
        }
    }
}

File: android/app/src/main/java/com/harvestglass/harvest/ui/chat/ChatViewModel.kt
package com.harvestglass.harvest.ui.chat

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.harvestglass.harvest.data.model.Message
import com.harvestglass.harvest.data.model.UserProfile
import com.harvestglass.harvest.data.service.ChatService
import com.harvestglass.harvest.data.model.SafetyAnalysis
import com.harvestglass.harvest.data.service.MatchService
import com.harvestglass.harvest.data.service.SafetyAnalysisService
import com.harvestglass.harvest.data.service.MindfulAnalysis
import com.harvestglass.harvest.data.service.MindfulMessagingService
import com.harvestglass.harvest.data.service.ProfileService
import com.harvestglass.harvest.ui.components.ReportTarget
import com.harvestglass.harvest.util.userMessage
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

data class ChatUiState(
    val messages: List<Message> = emptyList(),
    val partner: UserProfile? = null,
    val draft: String = "",
    val safetyAnalysis: SafetyAnalysis? = null,
    /** Shown as a banner above the transcript when the score drops below 50. */
    val safetyWarning: String? = null,
    val showReadyToMoveGate: Boolean = false,
    val isReadyToMove: Boolean = false,
    val readyToMoveReason: String? = null,
    /** Set once a share is recorded, so the screen can confirm it. */
    val readyToMoveMessage: String? = null,
    /** Set when the pre-send check flags a message; drives the warning sheet. */
    val mindfulWarning: MindfulAnalysis? = null,
    /** The flagged text, held so "Send Anyway" posts exactly what was typed. */
    val pendingSend: String? = null,
    /** The partner is typing right now. Clears itself after 3s of silence. */
    val isPartnerTyping: Boolean = false,
    val isLoading: Boolean = false,
    val isSending: Boolean = false,
    val error: String? = null
)

/**
 * Mirrors the message-transport half of Harvest/ViewModels/ChatViewModel.swift.
 *
 * The mindful pre-send warning and the safety "ready to move" analysis are NOT
 * ported here — both are OpenAI-backed and belong to the Gardener/AI subsystem.
 * Everything about sending, receiving, echo de-duplication and read receipts is
 * reproduced.
 */
@HiltViewModel
class ChatViewModel @Inject constructor(
    private val chatService: ChatService,
    private val profileService: ProfileService,
    private val matchService: MatchService,
    private val mindful: MindfulMessagingService,
    private val safety: SafetyAnalysisService
) : ViewModel() {

    private val _state = MutableStateFlow(ChatUiState())
    val state: StateFlow<ChatUiState> = _state.asStateFlow()

    private var conversationId: String = ""
    private var userId: String = ""
    private var messageJob: Job? = null
    private var typingJob: Job? = null
    private var typingSendJob: Job? = null
    private var typingDismissJob: Job? = null

    fun start(conversationId: String, userId: String, partnerUserId: String) {
        this.conversationId = conversationId
        this.userId = userId

        viewModelScope.launch {
            _state.update { it.copy(isLoading = true) }
            try {
                val messages = chatService.getMessages(conversationId)
                val partner = runCatching { profileService.getProfile(partnerUserId) }.getOrNull()
                _state.update { it.copy(messages = messages, partner = partner, error = null) }
            } catch (e: Exception) {
                _state.update { it.copy(error = e.userMessage()) }
            } finally {
                _state.update { it.copy(isLoading = false) }
            }

            messageJob?.cancel()
            messageJob = viewModelScope.launch {
                chatService.subscribeToMessages(conversationId).collect { msg ->
                    // Keyed on id: the sender already inserted it optimistically.
                    if (_state.value.messages.none { it.id == msg.id }) {
                        _state.update { it.copy(messages = it.messages + msg) }
                        // Only the other person's messages are scored — you are
                        // warned about your own before they are sent.
                        if (!msg.isSentBy(userId)) scoreIncoming(msg.content)
                    }
                }
            }

            loadSafetyAnalysis(partnerUserId)

            typingJob?.cancel()
            typingJob = viewModelScope.launch {
                chatService.subscribeToTyping(conversationId).collect { typistId ->
                    // Your own keystrokes come back on the same channel.
                    if (typistId == userId) return@collect

                    _state.update { it.copy(isPartnerTyping = true) }

                    // Typing has no "stopped" event, so the indicator expires
                    // on its own and every new keystroke restarts the clock.
                    typingDismissJob?.cancel()
                    typingDismissJob = viewModelScope.launch {
                        delay(TYPING_DISMISS_MS)
                        _state.update { it.copy(isPartnerTyping = false) }
                    }
                }
            }
        }
    }

    fun onDraftChange(text: String) {
        _state.update { it.copy(draft = text) }
        if (text.isNotEmpty()) scheduleTypingIndicator()
    }

    /**
     * Broadcasts one "typing" ping per burst of keystrokes rather than one per
     * character — the debounce is what keeps the channel quiet.
     */
    private fun scheduleTypingIndicator() {
        typingSendJob?.cancel()
        typingSendJob = viewModelScope.launch {
            delay(TYPING_DEBOUNCE_MS)
            if (conversationId.isEmpty() || userId.isEmpty()) return@launch
            runCatching { chatService.sendTypingIndicator(conversationId, userId) }
        }
    }

    /**
     * Runs the mindful check before sending. A flagged message opens the
     * warning instead — the send is deferred, not cancelled, so "Send Anyway"
     * still posts exactly what was typed.
     */
    fun send(content: String = _state.value.draft) = viewModelScope.launch {
        if (_state.value.isSending) return@launch
        val text = content.trim()
        if (text.isEmpty()) return@launch

        if (mindful.isEnabled) {
            val analysis = mindful.analyzeMessage(text)
            if (analysis.needsReview) {
                _state.update { it.copy(mindfulWarning = analysis, pendingSend = text) }
                return@launch
            }
        }

        deliver(text)
    }

    /** Dismisses the warning and puts the text back in the composer. */
    fun editFlaggedMessage() = _state.update {
        it.copy(mindfulWarning = null, pendingSend = null, draft = it.pendingSend ?: it.draft)
    }

    /** The nudge is not a block: this posts the message as typed. */
    fun sendAnyway() = viewModelScope.launch {
        val text = _state.value.pendingSend ?: return@launch
        _state.update { it.copy(mindfulWarning = null, pendingSend = null) }
        deliver(text)
    }

    private suspend fun deliver(text: String) {
        _state.update { it.copy(isSending = true, draft = "", error = null) }
        try {
            val sent = chatService.sendMessage(conversationId, userId, text)
            _state.update { s ->
                val alreadyThere = sent == null || s.messages.any { it.id == sent.id }
                s.copy(messages = if (alreadyThere) s.messages else s.messages + sent!!)
            }
        } catch (e: Exception) {
            // Hand the text back so a failed send isn't a lost message.
            _state.update { it.copy(draft = text, error = e.userMessage()) }
        } finally {
            _state.update { it.copy(isSending = false) }
        }
    }

    /**
     * Loads or creates the analysis for this conversation.
     *
     * Every failure here is silent: safety scoring is a background service, and
     * a chat that works is better than one blocked on it.
     */
    private suspend fun loadSafetyAnalysis(partnerUserId: String) {
        // A Seed conversation that never had a match has nothing to key on.
        val matchId = runCatching { chatService.matchId(conversationId) }.getOrNull() ?: return

        val analysis = runCatching {
            safety.getOrCreateAnalysis(
                matchId = matchId,
                userId = userId,
                otherUserId = partnerUserId
            )
        }.getOrNull() ?: return

        _state.update { it.copy(safetyAnalysis = analysis, safetyWarning = warningFor(analysis)) }
    }

    /** Below 50 the conversation gets a standing banner, as on iOS. */
    private fun warningFor(analysis: SafetyAnalysis): String? =
        if (analysis.safetyScore < 50) {
            "Safety concern detected. Be cautious in this conversation."
        } else {
            null
        }

    /** Scores one incoming message and refreshes the banner. */
    private suspend fun scoreIncoming(content: String?) {
        val analysis = _state.value.safetyAnalysis ?: return
        val text = content?.takeIf { it.isNotBlank() } ?: return

        runCatching { safety.analyzeMessage(text, analysis.id) }
        val refreshed = runCatching {
            safety.getOrCreateAnalysis(
                matchId = analysis.matchId,
                userId = userId,
                otherUserId = analysis.otherUserId
            )
        }.getOrNull() ?: return

        _state.update { it.copy(safetyAnalysis = refreshed, safetyWarning = warningFor(refreshed)) }
    }

    /** Opens the ready-to-move gate, refreshing its verdict first. */
    fun presentReadyToMoveGate() = viewModelScope.launch {
        val analysis = _state.value.safetyAnalysis
        if (analysis == null) {
            _state.update {
                it.copy(
                    showReadyToMoveGate = true,
                    isReadyToMove = false,
                    readyToMoveReason = "No safety analysis available yet."
                )
            }
            return@launch
        }

        val (ready, reason) = runCatching { safety.isReadyToMove(analysis.id) }
            .getOrDefault(false to "We couldn't check this conversation just now.")

        _state.update {
            it.copy(showReadyToMoveGate = true, isReadyToMove = ready, readyToMoveReason = reason)
        }
    }

    fun dismissReadyToMoveGate() = _state.update { it.copy(showReadyToMoveGate = false) }

    fun clearReadyToMoveMessage() = _state.update { it.copy(readyToMoveMessage = null) }

    /** Records that the user chose to share contact details. */
    fun markPreferredContactShared() = viewModelScope.launch {
        val analysis = _state.value.safetyAnalysis ?: return@launch
        if (!_state.value.isReadyToMove) return@launch

        try {
            safety.recordReadyToMoveDecision(
                userId = userId,
                matchId = analysis.matchId,
                conversationId = analysis.conversationId,
                safetyScore = analysis.safetyScore,
                approved = true,
                contactShared = true,
                contactMethod = "social"
            )
            _state.update {
                it.copy(
                    showReadyToMoveGate = false,
                    readyToMoveMessage =
                        "You're clear to share your preferred contact details in the chat."
                )
            }
        } catch (_: Exception) {
            _state.update {
                it.copy(
                    readyToMoveMessage =
                        "We couldn't record your sharing action. Please try again."
                )
            }
        }
    }

    /**
     * Which lexicon an incoming message trips, or null when it is clean or the
     * recipient has turned mindful messaging off.
     */
    fun flaggedCategory(content: String?): String? {
        if (!mindful.isEnabled) return null
        return mindful.localFlag(content.orEmpty())?.category
    }

    fun markRead(messageId: String) = viewModelScope.launch {
        runCatching { chatService.markAsRead(messageId) }
    }

    fun report(
        reportedUserId: String,
        category: String,
        description: String,
        target: ReportTarget = ReportTarget.Profile
    ) =
        viewModelScope.launch {
            try {
                matchService.reportUser(
                    reporterId = userId,
                    reportedUserId = reportedUserId,
                    category = category,
                    description = description,
                    targetType = target.typeString,
                    targetId = target.targetId
                )
                _state.update { it.copy(error = null) }
            } catch (e: Exception) {
                _state.update { it.copy(error = e.userMessage()) }
            }
        }

    fun block(blockedUserId: String) = viewModelScope.launch {
        try {
            matchService.blockUser(userId, blockedUserId)
            _state.update { it.copy(error = null) }
        } catch (e: Exception) {
            _state.update { it.copy(error = e.userMessage()) }
        }
    }

    fun unmatch(matchId: String) = viewModelScope.launch {
        try {
            matchService.unmatchUser(matchId)
            _state.update { it.copy(error = null) }
        } catch (e: Exception) {
            _state.update { it.copy(error = e.userMessage()) }
        }
    }

    override fun onCleared() {
        super.onCleared()
        messageJob?.cancel()
        typingJob?.cancel()
        typingSendJob?.cancel()
        typingDismissJob?.cancel()
    }

    companion object {
        /** One ping per burst of keystrokes, matching Swift's 500ms debounce. */
        private const val TYPING_DEBOUNCE_MS = 500L

        /** How long an indicator survives without another ping. */
        private const val TYPING_DISMISS_MS = 3000L
    }
}
